README.md
[![Build Status](https://travis-ci.org/BCS-io-provision/bcs_postgresql.svg?branch=master)](https://travis-ci.org/BCS-io-provision/bcs_postgresql)

## bcs_postgresql

This is a wrapper to give a company specific version for the [stock postgresql cookbook](https://github.com/sous-chefs/postgresql)

## Requirements

| Requirement | Tested Version                                                  |
| ----------- | ----------------------------------------------------------------|
| Chef        | 12.5+                                                           |
| Ruby        | 2.3.0+                                                          |
| Ubuntu      | 14.04+                                                          |

* [None US collation requires the locale wrapper, bcs_locale, recipie to have run.](https://github.com/BCS-io/bcs_locale)
* [sous-chefs postgresql cookbook](https://github.com/sous-chefs/postgresql)

## Usage

Most commonly included into other cookbooks and add a dependency in the metadata.
bcs_sets the locale to GB. This requires that the server is also GB and that postgres collation is set to be GB as well. bcs_locale, which sets locale to GB, should be in the runlist ahead of bcs_postgresql.

| File        | Command                                                         |
| ----------- | ----------------------------------------------------------------|
| berksfile   | cookbook 'postgresql', github: "sous-chefs/postgresql"       |
| berksfile   | cookbook 'bcs_postgresql', github: "BCS-io-provision/bcs_postgresql" |
| berksfile   | cookbook 'locale', github: "BCS-io/chef-locale"                 |
| berksfile   | cookbook 'bcs_locale', github: "BCS-io/bcs_locale"              |
| metadata.rb | depends 'locale'                                                |
| recipe/     | include_recipe 'bcs_postgresql'                                 |


Include in a runlist

````
    chef.run_list = [
      "recipe[bcs_locale::default]",
      "recipe[bcs_postgresql::default]"
  ]
````


## Attributes

### Usage

Password for the postgres user is automatically set if you're using chef server. For chef-solo you have to set this attribute.

`node['postgresql']['password']['postgresql']`


### Authentication

#### pg_hba

To control the security of Postgres means configuring pg_hba.

These security settings trust ALL internal connections. This means that ANY user on the system can log on as any database user, including the superuser postgres, without a password. The only user on the system will be the deployer. Your security requirements may vary.

Postgres Documentation on pg_hba http://www.postgresql.org/docs/9.2/static/auth-pg-hba-conf.html

````
default["postgresql"]["pg_hba"]  = [
  { type: "local", db: "all", user: "all",  addr: "", method: "trust"  },
  { type: "host",  db: "all", user: "all",  addr: "127.0.0.1/32", method: "trust"  },
  { type: "host",  db: "all", user: "all",  addr: "::1/128", method: "trust"  },
  { type: "host",  db: "all", user: "all",  addr: "0.0.0.0/0", method: "md5"  }
]
````

Do not include the default pg_hba if you are configuring your own pg_hba.

````
default["postgresql"]["pg_hba_defaults"]                 = false
````

### Default Postgres User

A typical customisation is to give the default Postgres superuser a password. You need postgres to authenticate the postgres user. You could do this by making this the first line in your pg_hba.

````
{ type: "local", db: "all", user: "postgres",  addr: "", method: "md5"  },
````

Then change the user postgres in the standard cookbook way.

### Locale

[The default is to take the collation of the operating system.](http://www.postgresql.org/docs/9.3/static/locale.html) This can be overriden just for postgres, however, I prefer to have the system set the locale so that every application has the same localisation setting.


## Recipes

The default runs the upstream `postgresql::libpq`, a Ubuntu requirement, `postgresql::server` the main database application, and the optional modules from `postgresql::contrib` - this is needed for array and hstore support.

### Testing

To test postgresql instance out: 

```
sudo psql -U postgres
```


## Author

Author:: Richard Wigley (richard.wigley@bcs.io)


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).
## Master

## [2.4.0] - 2018-05-14
### Changed
  - Prevent remote access
    - Not sure whey we added it - prefer to break something and fix
      than leave a large security question mark.
  - serverspec to inspec

## [2.3.0] - 2018-05-08
### Added
  - Support for Ubuntu 18.04

## [2.2.3] - 2018-13-31
### Changed
  - pinned sous-chef/postgresql to 6.1.1 - while they are making major changes to 
    the cookbook as they move it from recipies to library
  - updated Gems
  - works on Ruby 2.5.1


## [2.2.0] - 2017-04-24
### Changed
  - posgresql 5.1.0 => 5.2.0

## [2.1.0] - 2016-13-20
### Added
  - Rakefile added

### Changed
  - chef_solo => chef_zero
  - metadata version had been set to 2.0.4 when it should have been 2.0.0

## [2.0.0] - 2016-12-20
### Breaking Changes
  - Ubuntu 12.04 no longer supported
  - Chef 11 no longer supported

### Added
  - Ubuntu 16.04 is now supported

### Changed
  - Wrapped cookbook has been renamed from heavy-water to sous-chef 
  - Test Kitchen works on Mac

## [1.1.0] - 2015-08-18
### Changed
  - Update dependency on philper/postgresql from 0.15.0 to 0.16.0

### Added
  - Changelog


test/integration/default/default_test.rb
# # encoding: utf-8

# Inspec test for recipe letting::default

# The Inspec reference, with examples and extensive documentation, can be
# found at http://inspec.io/docs/reference/resources/

title 'bcs_postgresql::default'

def app_version(release = os[:release])
  if release == '14.04'
    '9.3'
  elsif release == '16.04' || release == '18.04'
    '9.5'
  end
end

describe package("postgresql-#{app_version}") do
  it { should be_installed }
end

if os[:release] == '14.04'
  describe service('postgresql') do
    it { should be_enabled  }
    it { should be_installed  }
    it { should be_running }
  end
elsif os[:release] == '16.04' || os[:release] == '18.04'
  describe systemd_service('postgresql') do
    it { should be_enabled }
    it { should be_installed }
    it { should be_running }
  end
end

describe file("/etc/postgresql/#{app_version}/main/pg_hba.conf") do
  it { should exist }
  its('content') { should match('83.67.10.5/32') }
end


test/integration/default/serverspec/default_spec.rb
# frozen_string_literal: true

require 'serverspec'

set :backend, :exec

describe 'bcs_postgresql::default' do
  it 'installed postgresql' do
    expect(package(app_version(os[:release]))).to be_installed
  end

  describe service 'postgresql' do
    if os[:release] == '14.04'
      # be_enabled was false - but moving away from 14 so not
      # investigating any further
      it { should be_running }
    elsif os[:release] == '16.04' || os[:release] == '18.04'
      it { should be_enabled }
      it { should be_running.under('systemd') }
    end
  end

  def app_version(os)
    if os == '14.04'
      'postgresql-9.3'
    elsif os == '16.04' || os == '18.04'
      'postgresql-9.5'
    end
  end
end


attributes/default.rb
# frozen_string_literal: true

#
# wrapper for sous-chefs/postgresql
#   - pinned at 6.1.1 while the move to lwrp
#   - we want to use a lwrp next as it is more precise than a general recipe
#     - alternative to sous-chef is https://github.com/express42/postgresql_lwrp

# Security Settings
default['postgresql']['pg_hba_defaults'] = false

# security applied when you try to make connection
default['postgresql']['pg_hba'] = [
  { type: 'local', db: 'all', user: 'all', addr: '', method: 'trust' },
  { type: 'host',  db: 'all', user: 'all', addr: '127.0.0.1/32', method: 'trust' },
  { type: 'host',  db: 'all', user: 'all', addr: '::1/128', method: 'trust' }
]

# chef-solo / chef-zero requires a password is set - chef-server sets password
#
default['postgresql']['password']['postgresql'] = ''

# IP security setting
#   - what ip addresses (interfaces really) will you listen on
default['postgresql']['config']['listen_addresses'] = 'localhost'
